Analytics/Integrations/Kahuna/SEGKahunaIntegration.m

```diff
@@ -38,42 +103,400 @@ - (void)validate {
+    id categoriesViewed = [existingAttributes valueForKey:KAHUNA_CATEGORIES_VIEWED];
+    if (categoriesViewed && [categoriesViewed isKindOfClass:[NSString class]]) {
+      NSMutableArray *aryOfCategoriesViewed = [[categoriesViewed componentsSeparatedByString:@","] mutableCopy];
+      if (![aryOfCategoriesViewed containsObject:value]) {
+        if (aryOfCategoriesViewed.count > 50) {
+          [aryOfCategoriesViewed removeObjectAtIndex:0];  // Remove the first object.
+        }
+        
+        [aryOfCategoriesViewed addObject:value];
+        [(*attributes) setValue:[aryOfCategoriesViewed componentsJoinedByString:@","] forKey:KAHUNA_CATEGORIES_VIEWED];
+      }
+    } else {
+      [(*attributes) setValue:value forKey:KAHUNA_CATEGORIES_VIEWED];
+    }
   } else {
-    // Although not documented, Kahuna wants revenue in cents
-    NSNumber *quantity = [properties objectForKey:@"quantity"] ?: @1;
-    [KahunaAnalytics trackEvent:event withCount:[quantity longValue] andValue:[revenue longValue] * 100];
+    // Since we do not have a category, we will store "none" for last view category and categories viewed list.
+    [(*attributes) setValue:KAHUNA_NONE forKey:KAHUNA_LAST_VIEWED_CATEGORY];
+    [(*attributes) setValue:KAHUNA_NONE forKey:KAHUNA_CATEGORIES_VIEWED];
+  }
+}
+
+- (void) addViewedProductElements:(NSMutableDictionary*__autoreleasing*) attributes fromProperties:(NSDictionary*) properties {
+  id kname = properties [KAHUNA_NAME];
+  if (KAHUNA_NOT_STRING_NULL_EMPTY(kname)) {
+    [(*attributes) setValue:kname forKey:KAHUNA_LAST_PRODUCT_VIEWED_NAME];
+  }
+  
+  [self addViewedProductCategoryElements:attributes fromProperties:properties];
+}
+
+- (void) addAddedProductElements:(NSMutableDictionary*__autoreleasing*) attributes fromProperties:(NSDictionary*) properties {
+  id kname = properties [KAHUNA_NAME];
+  if (KAHUNA_NOT_STRING_NULL_EMPTY(kname)) {
+    [(*attributes) setValue:kname forKey:KAHUNA_LAST_PRODUCT_ADDED_TO_CART_NAME];
+  }
+  
+  id category = properties [KAHUNA_CATEGORY];
+  if (!KAHUNA_NOT_STRING_NULL_EMPTY(category)) {
+    category = KAHUNA_NONE;
+  }
+  
+  [(*attributes) setValue:category forKey:KAHUNA_LAST_PRODUCT_ADDED_TO_CART_CATEGORY];
+}
+
+- (void) addCompletedOrderElements:(NSMutableDictionary*__autoreleasing*) attributes fromProperties:(NSDictionary*) properties {
+  id discount = properties [KAHUNA_DISCOUNT];
+  if ([discount isKindOfClass:[NSString class]] || [discount isKindOfClass:[NSNumber class]]) {
+    [(*attributes) setValue:discount forKey:KAHUNA_LAST_PURCHASE_DISCOUNT];
+  } else {
+    [(*attributes) setValue:@0 forKey:KAHUNA_LAST_PURCHASE_DISCOUNT];
   }
 }
 
 - (void)screen:(NSString *)screenTitle properties:(NSDictionary *)properties options:(NSDictionary *)options {
-  // Track the screen view as an event.
-  [self track:SEGEventNameForScreenTitle(screenTitle) properties:properties options:options];
+  id trackAllPages = [self.settings objectForKey:@"trackAllPages"];
```

Comment:
The extra checks have been added to ensure "trackAllPages" is indeed a NSNumber class and not something else that can possibly thrown an exception when booleanValue is called on the object.